### PR TITLE
Implement Application Default Credentials for the google connector

### LINKED ISSUE #1756: Support workload identity in Google connector
Workload identity accessing GoogleAPIs without having a static service account token. GKE metadata server available in each pod is used to fetch tokens.

Workload identity tokens work as a replacement for GSA token, they are not accepted by GSuite Admin API, because Admin API requires impersonalization for GSuite admin. Impersonalization IS possible with workload identity token, hence this proposal.

Example use of GSuite admin impersonalization can be found here: https://github.com/salrashid123/oauth2#impersonated-credentials-with-domain-wide-delegation

So proposal is following:
 - if google connector config specifies `adminEmail`, but does not specifies `serviceAccountFilePath` Google connector attempts to use workload identity token to impersonate GSuite Admin and create Admin API client instance

### PR DESCRIPTION
**Pull request by ichbinfrog**
#### Overview

Implements [workload identity](https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity) to the google connector. 

#### What this PR does / why we need it

Closes #1756 by implementing workload identity for the google connector. 

This PR has been tested by:
- Building the dex image
- Injecting the image into the Argocd helm chart with Google Groups integration (see related [doc](https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/google/#openid-connect-plus-google-groups-using-dex))
- Testing if Groups are visible within Argocd

I also wanted to add unit test but it required some changes to the `Open` interface to add  `option.ClientOptions...` or `interface{}...`.

Another similar [PR](https://github.com/dexidp/dex/pull/2122) exists but testing it yields an `404: Domain not found` error.

#### Does this PR introduce a user-facing change?

```release-note
Add application default credentials for the google workspace connector
```

### CODE REVIEW

**Comment by nabokihms** [comment]
@ichbinfrog sorry for the delay with this. The idea looks great! I only wonder about testing, adding a test is still desired. Besides it, lgtm.

File: connector/google/google.go [hunk 1]
```diff
@@ -279,37 +279,36 @@ func (c *googleConnector) getGroups(email string, fetchTransitiveGroupMembership
...
-
-	srv, err := admin.NewService(ctx, option.WithHTTPClient(client))
+	if serviceAccountFilePath == "" {
+		credential, err := google.FindDefaultCredentials(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch application default credentials: %w", err)
+		}
+		jsonCredentials = credential.JSON
+	} else {
+		jsonCredentials, err = os.ReadFile(serviceAccountFilePath)
```
Comment (nabokihms) [inline connector/google/google.go:301]: Maybe we should log a warning, that default credentials file is used.

**Comment by nabokihms** [comment]
Moving this one to the v2.34 milestone.

**Comment by jinnjwu** [comment]
our dex is hosting in gke, we need this feature too

**Comment by ichbinfrog** [comment]
Thanks for the review @nabokihms and sorry for the delay, I've added the warning log as well as some unit tests for the logic. 

Regarding the unit test implementation, I've given it a try. It did however require that the `Open` interface become to allow creating the Google Directory Service pointing to a local endpoint:

```golang
Open(id string, logger log.Logger, opts ...interface{}) (connector.Connector, error)
```

In the long term, this change would allow us to add further unit testing of the google connector implementation by mocking the API similarly to what's currently in the oauth connector.

File: connector/google/google.go [hunk 2]
```diff
@@ -279,37 +289,37 @@ func (c *googleConnector) getGroups(email string, fetchTransitiveGroupMembership
...
-	config.Subject = email
+	var jsonCredentials []byte
+	var err error
 
 	ctx := context.Background()
-	client := config.Client(ctx)
-
-	srv, err := admin.NewService(ctx, option.WithHTTPClient(client))
+	if serviceAccountFilePath == "" {
+		logger.Warn("the application default credential is used since the service acocunt file path is not used")
```
Comment (shuker85) [inline connector/google/google.go:305]: small typo on `account`

File: connector/google/google_test.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	// 	json.NewEncoder(w).Encode(&admin.Groups{
+	// 		Groups: []*admin.Group{},
+	// 	})
+	// })
+	return httptest.NewServer(mux)
+}
+
+func newConnector(config *Config, serverURL string) (*googleConnector, error) {
+	log := logrus.New()
+	conn, err := config.Open("id", log, option.WithEndpoint(serverURL))
```
Comment (nabokihms) [inline connector/google/google_test.go:30]: If only endpoint changes are required for now, I'd suggest making it through the private connector property avoiding global changes.

**Review by nabokihms — CHANGES_REQUESTED** [review]
Looks great! I have only one convern left. Despite it, everything is ok.

File: connector/google/google_test.go (hunk 3 above)
Comment (ichbinfrog) [reply connector/google/google_test.go:30]: Agreed, in hindsight, the behavior to be tested in this PR does not even require that option.

**Comment by ichbinfrog** [comment]
There's no problem regarding the delay. My PR updates are also extremely delayed 😅

**Review by nabokihms — APPROVED** [review]
@ichbinfrog thanks! LGTM.

**Pull request merged**